[AMDGPU] Enable vectorization of i8 values.

doru1004 opened the pull request:
This patch adjusts the cost model to account for the ability of the AMDGPU optimizer to group together i8 values into i32 values.

doru1004 commented on llvm/test/CodeGen/AMDGPU/vectorize-i8-as-i32.ll:
```diff
@@ -0,0 +1,88 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt -S -mtriple=amdgcn-amd-amdhsa -mcpu=hawaii -passes=slp-vectorizer %s | FileCheck -check-prefixes=GFX7 %s
```
Good catch thank you ! :)

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -12971,9 +12971,16 @@ BoUpSLP::getEntryCost(const TreeEntry *E, ArrayRef<Value *> VectorizedVals,
-              LI0->getPointerAddressSpace(), CostKind, TTI::OperandValueInfo());
+          if (VecTy->getElementType() ==
+                  IntegerType::getInt8Ty(VecTy->getContext()) &&
+              TTI->canVectorizei8s()) {
```
Bad check, this should be here

doru1004 replied:
Fixed

alexey-bataev reviewed (changes requested):
The logic in the SLP vectorizer should be target-independent

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -20927,7 +20934,11 @@ bool SLPVectorizerPass::tryToVectorizeList(ArrayRef<Value *> VL, BoUpSLP &R,
     // code during codegen).
     auto *VecTy = getWidenedType(ScalarTy, VF);
-    if (TTI->getNumberOfParts(VecTy) == VF)
+    unsigned NumParts = TTI->getNumberOfParts(VecTy);
+    if (TTI->canVectorizei8s() &&
+        VecTy->getElementType() == IntegerType::getInt8Ty(VecTy->getContext()))
+      NumParts = 1;
+    if (NumParts == VF)
```
Same here, it should not be in SLP vectorizer, it should be the part of getNumberOfParts logic in GCNTTIImpl

doru1004 replied:
Fixed

jrbyrnes commented on llvm/lib/Analysis/TargetTransformInfo.cpp:
```diff
@@ -1419,6 +1419,10 @@ unsigned TargetTransformInfo::getMaxNumArgs() const {
   return TTIImpl->getMaxNumArgs();
 }
 
+bool TargetTransformInfo::canVectorizei8s() const {
```
Not sure about this TTI hook. In general, there is no special about i8s. They just happen to require certain special handling on AMDGPUs. However, for other targets, i8s may be first class citizens / legal, so they are already legalized by default -- x86 for example I believe has legal i8s

doru1004 replied:
Right, I think the name of that hook can be improved. The word vectorize in there is not quite correct.

jrbyrnes commented on llvm/test/Analysis/CostModel/AMDGPU/shufflevector.ll:
```diff
@@ -396,157 +396,157 @@ define amdgpu_kernel void @shufflevector_i16(<2 x i16> %vec1, <2 x i16> %vec2) {
+; ALL-SIZE-NEXT:  Cost Model: Found an estimated cost of 0 for instruction: %shuf03 = shufflevector <2 x i8> %vec1, <2 x i8> %vec1, <2 x i32> <i32 0, i32 3>
+; ALL-SIZE-NEXT:  Cost Model: Found an estimated cost of 0 for instruction: %shuf30 = shufflevector <2 x i8> %vec1, <2 x i8> %vec1, <2 x i32> <i32 3, i32 0>
+; ALL-SIZE-NEXT:  Cost Model: Found an estimated cost of 0 for instruction: %shuf33 = shufflevector <2 x i8> %vec1, <2 x i8> %vec1, <2 x i32> <i32 3, i32 3>
+; ALL-SIZE-NEXT:  Cost Model: Found an estimated cost of 0 for instruction: %shuf12 = shufflevector <2 x i8> %vec1, <2 x i8> %vec1, <2 x i32> <i32 1, i32 2>
```
Something has happened to make these cost calculations incorrect.

Currently, the cost for this particular shufflevector is 4 which is probably not very accurate. But a cost of 0 is certainly wrong -- 0 cost implies that the IR instruction can be optimized away / does not map to any instruciton in the assembly.

https://godbolt.org/z/7zPds31eM

Here we see we need multiple instructions to produce this shufflevector: 1.  `v_mov_b32_e32 v0, 0xc0c0001` to produce the mask, 2. `v_perm_b32 v0, 0, s0, v0` to do the shufflevector. Usually we would need a third to truncate the result of the perm to 16 bits (due to the shufflevector being v2i8), but the user (ds_write) is able to select the lower 16 bits without a truncate.

doru1004 replied:
No longer 0, I'll mark this as fixed but can revisit in light of the latest changes.

shiltian commented on llvm/test/Transforms/SLPVectorizer/AMDGPU/vectorize-i8-as-i32.ll:
```diff
@@ -0,0 +1,88 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt -S -mtriple=amdgcn-amd-amdhsa -mcpu=hawaii -passes=slp-vectorizer %s | FileCheck -check-prefixes=GFX7 %s
+; RUN: opt -S -mtriple=amdgcn-amd-amdhsa -mcpu=fiji -passes=slp-vectorizer %s | FileCheck -check-prefixes=GFX8PLUS,GFX8 %s
+; RUN: opt -S -mtriple=amdgcn-amd-amdhsa -mcpu=gfx900 -passes=slp-vectorizer %s | FileCheck -check-prefixes=GFX8PLUS,GFX9 %s
```
The script can't merge check lines (see the last lines of the file), so it'd be fine to just use `--check-prefix`.

shiltian commented on llvm/test/Transforms/SLPVectorizer/AMDGPU/vectorize-i8-as-i32.ll:
```diff
@@ -0,0 +1,88 @@
+  store <16 x i8> %vec03, ptr %out1
+  ret void
+}
+
+
+
+
```
Remove those many empty lines and just leave one at the end.

shiltian commented on llvm/include/llvm/Analysis/TargetTransformInfo.h:
```diff
@@ -1907,6 +1907,10 @@ class TargetTransformInfo {
 
+  /// \return Returns true if vectorizing 4 x i8s into an i32 is possible.
+  /// Currently only used by the SLP vectorizer.
+  bool canVectorizei8s() const;
```
IMHO this is not really a good function names, especially the `i8s` part.

doru1004 replied:
It's been removed.

shiltian commented on llvm/include/llvm/Analysis/TargetTransformInfo.h:
```diff
@@ -1907,6 +1907,10 @@ class TargetTransformInfo {
   /// pad to. Default is no padding.
   unsigned getNumBytesToPadGlobalArray(unsigned Size, Type *ArrayType) const;
 
+  /// \return Returns true if vectorizing 4 x i8s into an i32 is possible.
```
```suggestion
  /// \return true if vectorizing 4 x i8s into an i32 is possible.
```

doru1004 replied:
Function was removed.

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:
```diff
@@ -1423,3 +1430,25 @@ void GCNTTIImpl::collectKernelLaunchBounds(
+      VecTy->getElementType() == IntegerType::getInt8Ty(VecTy->getContext()))
+    return 1;
+  return BaseT::getNumberOfParts(Tp);
+}
```
Missing empty line at the end

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.h:
```diff
@@ -282,6 +282,16 @@ class GCNTTIImpl final : public BasicTTIImplBase<GCNTTIImpl> {
+      const Instruction *I = nullptr);
+
+  /// \return numer of parts in this type.
+  unsigned getNumberOfParts(Type *Tp);
```
What do you mean by "parts"?

jrbyrnes commented on llvm/test/Analysis/CostModel/AMDGPU/add-sub.ll:
```diff
@@ -126,24 +126,24 @@ define amdgpu_kernel void @add_i16() #0 {
+; ALL-NEXT:  Cost Model: Found an estimated cost of 1 for instruction: %v4i8 = add <4 x i8> undef, undef
+; ALL-NEXT:  Cost Model: Found an estimated cost of 2 for instruction: %v5i8 = add <5 x i8> undef, undef
+; ALL-NEXT:  Cost Model: Found an estimated cost of 2 for instruction: %v6i8 = add <6 x i8> undef, undef
+; ALL-NEXT:  Cost Model: Found an estimated cost of 8 for instruction: %v32i8 = add <32 x i8> undef, undef
```
The cost of v32i8 add should be 32 

https://godbolt.org/z/o5W1Y1oer

In the ISA, we don't have an instruction to do vectorized i8 adds. In the example, we see that the v32i8 vector add is scalarized into a bunch of individual i8 adds. Given that we need 32 instructinos, the cost of the v32i8 add should be 32.

The result of changing the cost to 8 is that the vector version has a cost savings of 24 over the scalar version. Thus, for a vectorizeable chain, the vectorized version can degrade relative to the scalar version by a cost of up to 23. Since the cost savings of 24 doesn't actually translate to improved assembly, this means we will allow code degradations for no benefit.

doru1004 replied:
Are you expecting the following code to be worth vectorizing from a cost model perspective?
```
  %el0 = extractelement <16 x i8> %invec, i64  0
   %el1 = extractelement <16 x i8> %invec, i64  1
   %el2 = extractelement <16 x i8> %invec, i64  2
   %el3 = extractelement <16 x i8> %invec, i64  3
   %mul0 = mul i8 %el0, 1
   %mul1 = mul i8 %el1, 1
   %mul2 = mul i8 %el2, 1
   %mul3 = mul i8 %el3, 1
   %add0 = add i8 %mul0, 1
   %add1 = add i8 %mul1, 1
   %add2 = add i8 %mul2, 1
   %add3 = add i8 %mul3, 1
   %vecins0 = insertelement <16 x i8> poison, i8 %add0, i64 0
   %vecins1 = insertelement <16 x i8> %vecins0, i8 %add1, i64 1
   %vecins2 = insertelement <16 x i8> %vecins1, i8 %add2, i64 2
   %vecins3 = insertelement <16 x i8> %vecins2, i8 %add3, i64 3
   store <16 x i8> %vecins3, ptr %out
   ret void
```

Because with the patch it can currently get transformed to this:

```
  %0 = shufflevector <16 x i8> %invec, <16 x i8> poison, <4 x i32> <i32 0, i32 1, i32 2, i32 3>
  %1 = mul <4 x i8> %0, splat (i8 1)
  %2 = add <4 x i8> %1, splat (i8 1)
  %3 = shufflevector <4 x i8> %2, <4 x i8> poison, <16 x i32> <i32 0, i32 1, i32 2, i32 3, i32 poison, i32 poison, i32 poison, i32 poison, i32 poison, i32 poison, i32 poison, i32 poison, i32 poison, i32 poison, i32 poison, i32 poison>
  store <16 x i8> %3, ptr %out, align 16
  ret void
```

So there is no benefit to having this vectorized code?

doru1004 replied:
Ok I just verified that the assembly for the 2 cases above is more or less identical in which case the current cost model is already correct in that regard.

doru1004 commented:
ping

jrbyrnes commented on llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:
```diff
@@ -1423,3 +1425,30 @@ void GCNTTIImpl::collectKernelLaunchBounds(
+        VecTy->getElementType() ==
+            IntegerType::getInt8Ty(VecTy->getContext())) {
+      unsigned ElementCount = VecTy->getElementCount().getFixedValue();
+      return ((ElementCount - 1) / 4) + 1;
```
It shouldn't be the ratio of i32 / i8 -- but rather the amount of loads needed. I think, it should always be 1 as long as the width of the load is supported by hardware. May want to asset / check the vector width against the supported load/store width for the address space.

doru1004 replied:
Updated the formula to take into consideration the load/store register bit width.

jrbyrnes commented on llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:
```diff
@@ -562,6 +563,7 @@ InstructionCost GCNTTIImpl::getArithmeticInstrCost(
     if (ST->has16BitInsts() && SLT == MVT::i16)
       NElts = (NElts + 1) / 2;
 
+    // i32
```
Accidental comment? Should be full sentence.

jrbyrnes commented on llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:
```diff
@@ -344,9 +344,10 @@ unsigned GCNTTIImpl::getMinVectorRegisterBitWidth() const {
-  return (ElemWidth == 16 && ST->has16BitInsts()) ? 2
-       : (ElemWidth == 32 && ST->hasPackedFP32Ops()) ? 2
-       : 1;
+  return ElemWidth == 8                                ? 4
```
Do all the TTI->get*Cost queries handled i8/vector i8 types?

jrbyrnes replied:
If we allow SLP to consider i8s for vectorization, then we are relying on the cost model to make the right decision of vectorizing certain sequcnes vs not. Given that we are relying on the cost model now to make accurate decisions, we should double check that the other cost queries are behaving as expected for i8s / vector i8s.

Since these cost queries are used elsewhere, I assume they already handle these types well, but it's worth double checking.

doru1004 replied:
This is only called in 2 places so I added an extra test to cover that.

jrbyrnes replied:
The TTI::get*Cost queries are called in many places. Can you please double check that the ones we have implemented in AMDGPUTargetTransformInfo either revert to the base handling or are explicitly customized for i8

doru1004 replied:
Why the cost queries? This function is only called in a handful of places and is not called by all cost queries?

jrbyrnes replied:
By changing the `getMaximimumVF` and `getNumberOfParts` we are effectively enabling SLPVectorization for i8s. Thus, the control for i8 vectorization has shifted from this function (and `getNumberOfParts`) to the cost queries. If the cost queries say the vector cost is lower, then we will vectorize. So, since we are transferring control, we should be sure that it was working as expected.

doru1004 replied:
I checked and no other arch specific changes (other than the ones in this patch) are affected by the changes to these functions for i8s.

jrbyrnes commented on llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:
```diff
@@ -344,9 +344,10 @@ unsigned GCNTTIImpl::getMinVectorRegisterBitWidth() const {
-       : (ElemWidth == 32 && ST->hasPackedFP32Ops()) ? 2
-       : 1;
+  return ElemWidth == 8                                ? 4
+         : (ElemWidth == 16 && ST->has16BitInsts())    ? 2
```
It doesn't seem right that we will try to vectorize i8s in all cases but not for i16s.

doru1004 replied:
Yes I wondered this myself. It seems that this should be unconditional as well. I don't know (because of lack of comments) what the original intent was so I left it untouched as this seems a change which is unrelated to this patch.

doru1004 replied:
I would like to keep this patch about i8s only. If we want to relax conditions for other types it can be done in a separate PR.

jrbyrnes reviewed (commented):
If we are going to have a custom cost model for memory instructions we should probably add a test to test/Analysis/CostModel/AMDGPU

doru1004 commented:
> If we are going to have a custom cost model for memory instructions we should probably add a test to test/Analysis/CostModel/AMDGPU

Also added.

jrbyrnes commented on llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:
```diff
@@ -1423,3 +1424,32 @@ void GCNTTIImpl::collectKernelLaunchBounds(
+        VecTy->getElementType() ==
+            IntegerType::getInt8Ty(VecTy->getContext())) {
+      unsigned ElementCount = VecTy->getElementCount().getFixedValue();
+      return (8 * (ElementCount - 1) /
```
`return (8 * (ElementCount - 1) / getLoadStoreVecRegBitWidth(AddressSpace)) + 1;`
->
`return (DL.getTypeSizeInBits(VecTy) / getLoadStoreVecRegBitWidth(AddressSpace)) + 1;`

jrbyrnes commented on llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:
```diff
@@ -1423,3 +1424,32 @@ void GCNTTIImpl::collectKernelLaunchBounds(
+                                            TTI::TargetCostKind CostKind,
+                                            TTI::OperandValueInfo OpInfo,
+                                            const Instruction *I) {
+  if (VectorType *VecTy = dyn_cast<VectorType>(Src))
```
I think the cost should be the same for any illegal type -- so instead of checking for i8 vectors can check for illegal types.

```
 TargetLoweringBase::LegalizeKind LK = TLI->getTypeConversion(Src->getContext(), EVT::getEVT(Src->getScalarType(), false));
 if (LK.first != TargetLoweringBase::TypeLegal) { }
```

doru1004 replied:
I would like to keep this patch about i8s only with other types potentially enabled by future patches.

jrbyrnes commented on llvm/test/Analysis/CostModel/AMDGPU/load_i8.ll:
```diff
@@ -0,0 +1,466 @@
+define void @loads_align4(i32 %arg) {
+  ; Scalars
+; GFX90A-LABEL: 'loads_align4'
+; GFX90A-NEXT:  Cost Model: Found an estimated cost of 1 for instruction: %1 = load i8, ptr poison, align 4
```
Can you add tests for different types too (e.g. i1 and i16) and different addressspaces.

alexey-bataev commented:
No more comments from my side, since all changes are in TTI, not a vectorizer

doru1004 commented:
> No more comments from my side, since all changes are in TTI, not a vectorizer

Thank you Alexey.

arsenm reviewed (commented):
Can you fix the description to be more specific about what it enables vectorization of? Is this only for loads and stores? We already have the load/store vectorizer to mostly ignore the cost of vector operations. Most of the tests look like no-ops, can you split this to have a pre-commit adding the baseline tests?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:
```diff
@@ -344,9 +344,10 @@ unsigned GCNTTIImpl::getMinVectorRegisterBitWidth() const {
-  return (ElemWidth == 16 && ST->has16BitInsts()) ? 2
-       : (ElemWidth == 32 && ST->hasPackedFP32Ops()) ? 2
-       : 1;
+  return ElemWidth == 8                                ? 4
```
And should this be filtering on the opcode? Is this only enabling phi?

doru1004 replied:
No filtering is needed.

doru1004 commented:
> Can you fix the description to be more specific about what it enables vectorization of? Is this only for loads and stores? We already have the load/store vectorizer to mostly ignore the cost of vector operations. Most of the tests look like no-ops, can you split this to have a pre-commit adding the baseline tests?

This is not particular to loads and stores. This is about letting the cost model decide when it is profitable to group together i8s into an i32 instruction. Some instructions are not deemed profitable, hence the tests that seem to not be doing anything but in some cases it is. It is a particularly sensitive issue for load/store instructions hence maybe the extra focus on that but otherwise it is not limited to one type of instruction.

doru1004 commented:
I am in the process of precommitting the tests: https://github.com/llvm/llvm-project/pull/138801 so if someone can approve that PR I can merge and then rebase this patch so changes are more obvious.

jrbyrnes commented on llvm/test/Transforms/SLPVectorizer/AMDGPU/vectorize-i8-as-i32.ll:
```diff
@@ -0,0 +1,1096 @@
+; RUN: opt -S -mtriple=amdgcn-amd-amdhsa -mcpu=fiji -passes=slp-vectorizer %s | FileCheck -check-prefix=GFX8 %s
+; RUN: opt -S -mtriple=amdgcn-amd-amdhsa -mcpu=gfx900 -passes=slp-vectorizer %s | FileCheck -check-prefix=GFX9 %s
+
+define protected amdgpu_kernel void @arith_1(<16 x i8> %invec, ptr %out, i32 %flag) {
```
I think these arith_* tests are negative tests for enabling i8 vectorization, but this variant should be deleted as there are not multiple chains

jrbyrnes commented on llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:
```diff
@@ -344,9 +344,12 @@ unsigned GCNTTIImpl::getMinVectorRegisterBitWidth() const {
-       : 1;
+  // For a given width return the max 0number of elements that can be combined
+  // into a wider bit value:
+  return ElemWidth == 8                                ? 4
```
I think we should add `ST->isGFX8Plus()` / `ST->getGeneration() >= AMDGPUSubtarget::VOLCANIC_ISLANDS` -- v_perm is only available on these architectures and is the reason why vectorized shuffles have a cost benefit

jrbyrnes reviewed (commented):
LGTM but please wait for @arsenm for approval.

The one outstanding concern that we discussed offline (I think) is that this method of vectorizing does not encourage vectorization of other illegal types that may benefit from cross BB type coercion. We don't really have any strong present need that I know of, but we may encounter in the future. But, it is probably best to address such a need as it occurs rather than spending a bunch of time trying to future proof.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:
```diff
@@ -344,9 +344,13 @@ unsigned GCNTTIImpl::getMinVectorRegisterBitWidth() const {
-       : 1;
+  // For a given width return the max 0number of elements that can be combined
+  // into a wider bit value:
+  bool isGFX8Plus = ST->getGeneration() >= AMDGPUSubtarget::VOLCANIC_ISLANDS;
```
Don't do generation checks. This is already identical to has16BitInsts

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:
```diff
@@ -344,9 +344,12 @@ unsigned GCNTTIImpl::getMinVectorRegisterBitWidth() const {
-       : 1;
+  // For a given width return the max 0number of elements that can be combined
+  // into a wider bit value:
+  return (ElemWidth == 8 && ST->has16BitInsts())       ? 4
```
```suggestion
  return (ElemWidth == 8)       ? 4
```

Don't see has16Bits is really relevant to the 8-bit case. I guess hasSDWA would be a more plausible reason, but given this is artificial usage anyway it probably can be unconditional.

doru1004 replied:
This is what you had suggested previously but I'm happy to remove it.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:
```diff
@@ -1443,3 +1447,31 @@ void GCNTTIImpl::collectKernelLaunchBounds(
+  if (VectorType *VecTy = dyn_cast<VectorType>(Src)) {
+    if ((Opcode == Instruction::Load || Opcode == Instruction::Store) &&
+        VecTy->getElementType()->isIntegerTy(8)) {
+      return ((DL.getTypeSizeInBits(VecTy) - 1) /
+              getLoadStoreVecRegBitWidth(AddressSpace)) +
+             1;
```
This is treating each load/store as a cost of 1. We ought to be treating memory operations as more expensive? We weren't already overriding those costs?

doru1004 replied:
This is how other load/stores are being treated for vectorization purposes (even for other type widths) otherwise they would not be vectorized.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:
```diff
@@ -1443,3 +1447,31 @@ void GCNTTIImpl::collectKernelLaunchBounds(
+  if (VectorType *VecTy = dyn_cast<VectorType>(Tp)) {
+    if (VecTy->getElementType()->isIntegerTy(8)) {
+      unsigned ElementCount = VecTy->getElementCount().getFixedValue();
+      return ((ElementCount - 1) / 4) + 1;
```
Use divideCeil?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:
```diff
@@ -1443,3 +1447,31 @@ void GCNTTIImpl::collectKernelLaunchBounds(
+  if (VectorType *VecTy = dyn_cast<VectorType>(Src)) {
+    if ((Opcode == Instruction::Load || Opcode == Instruction::Store) &&
+        VecTy->getElementType()->isIntegerTy(8)) {
+      return ((DL.getTypeSizeInBits(VecTy) - 1) /
```
Use divideCeil?

arsenm reviewed (approved):
LGTM with nits. I think the load/store costs need to be revisited, we're treating them as as cheap as cheap ALU instructions which they certainly are not (also ignoring unaligned cost)

doru1004 commented:
> LGTM with nits. I think the load/store costs need to be revisited, we're treating them as as cheap as cheap ALU instructions which they certainly are not (also ignoring unaligned cost)

Since this is a problem which affects all load/store instructions regardless of type we will address this in a future PR to maintain consistency.

erichkeane committed:
[OpenACC][Docs] Add a release note for Clang 21 (#145938)

This patch adds a release note that explains the current status of
OpenACC in Clang. Currently we cannot actually make an executable
because the OpenACC dialect of MLIR doesn't support any amount of
lowering to LLVM-IR, so the usefulness of OpenACC is entirely for
front-end related uses, such as tooling or semantic checking.
